lib/spack/spack/environment/depfile.py

```diff
@@ -0,0 +1,242 @@
+        self.all_pkg_identifiers: List[str] = []
+
+        # All install and install-deps targets
+        self.all_install_related_targets: List[str] = []
+
+        # Convenience shortcuts: ensure that `make install/pkg-version-hash` triggers
+        # <absolute path to env>/.spack-env/makedeps/install/pkg-version-hash in case
+        # we don't have a custom make target prefix.
+        self.phony_convenience_targets: List[str] = []
+
+        for node in adjacency_list:
+            tgt = self._safe_name(node.target)
+            self.all_pkg_identifiers.append(tgt)
+            self.all_install_related_targets.append(self._install_target(tgt))
+            self.all_install_related_targets.append(self._install_deps_target(tgt))
+            if make_prefix is None:
+                self.phony_convenience_targets.append(os.path.join("install", tgt))
+                self.phony_convenience_targets.append(os.path.join("install-deps", tgt))
+
+    def _safe_name(self, spec: spack.spec.Spec) -> str:
+        return spec.format("{name}-{version}-{hash}")
+
+    def _target(self, name: str) -> str:
+        # The `all` and `clean` targets are phony. It doesn't make sense to
+        # have /abs/path/to/env/metadir/{all,clean} targets. But it *does* make
+        # sense to have a prefix like `env/all`, `env/clean` when they are
+        # supposed to be included
+        if name in ("all", "clean") and os.path.isabs(self.make_prefix):
+            return name
+        else:
+            return os.path.join(self.make_prefix, name)
+
+    def _install_target(self, name: str) -> str:
+        return os.path.join(self.make_prefix, "install", name)
+
+    def _install_deps_target(self, name: str) -> str:
+        return os.path.join(self.make_prefix, "install-deps", name)
+
+    def to_dict(self):
+        return {
+            "all_target": self._target("all"),
+            "env_target": self._target("env"),
+            "clean_target": self._target("clean"),
+            "all_install_related_targets": " ".join(self.all_install_related_targets),
+            "root_install_targets": " ".join(self.root_install_targets),
+            "dirs_target": self._target("dirs"),
+            "environment": self.env_path,
+            "install_target": self._target("install"),
+            "install_deps_target": self._target("install-deps"),
+            "any_hash_target": self._target("%"),
+            "jobserver_support": self.jobserver_support,
+            "adjacency_list": self.make_adjacency_list,
+            "phony_convenience_targets": " ".join(self.phony_convenience_targets),
+            "pkg_ids_variable": self.pkg_identifier_variable,
+            "pkg_ids": " ".join(self.all_pkg_identifiers),
+        }
+
+    @staticmethod
+    def from_env(
+        env: ev.Environment,
+        filter_specs: Optional[List[spack.spec.Spec]],
+        pkg_buildcache: str,
+        dep_buildcache: str,
+        make_prefix: Optional[str],
+        jobserver: bool,
```

Comment:
Can we make these keyword only arguments, with default values?

Reply:
Hm, not sure if it makes sense to provide default values, they're kinda mapped from command args, and with defaults it's easier to miss one.

Reply:
I was thinking of using this internally in core Spack, and requiring 6 args on each call is really a lot.

Reply:
As an alternative, can we gather these in a configuration struct that needs to be passed explicitly, and have a factory returning a default configuration object? In commands we can construct the configuration object explicitly, so we won't miss any new option, in other places in the code we could:
```python
model = MakefileModel.from_env(env, configuration=default_makefile_configuration())
```

Reply:
Missed your message, but feel free to make that change ^ in a PR once you know how you would use this class